Feature/se 1381.add access token env var support

slemasne opened the pull request:
# Pull Request Checklist

- [x] Read the [contributing guidelines](../docs/CONTRIBUTING.md)
- [x] Tests pass
- [x] Raised the PR against the `develop` branch

# Description of the PR

* The objective of this PR - allow users to pass in a PAT as an env variable.
* The env vars should only be used when the user does not explicitly pass a secret file or token as a param
* If, however, the user sets ApiFactory(token=None), then the logic should fall back to the env pat. This is need to support the sample-notbooks RefreshingToken() workflow. 

## High-level workflow for PAT as an env var

From a high-level, the new workflow is as follows:

1. Initialize the `ApiFactory` with no params but with a PAT in the env vars
2. The `ApiFactory` uses an `ApiClientBuilder` which loads configuration from the `ApiConfigurationLoader`
3. The `ApiConfigurationLoader`  loads the PAT from the env vars and uses the PAT for authentication (if a token has not been passed as a param directly)

rizwansaeed commented on sdk/tests/utilities/test_api_client_factory_auth_flow.py:
```diff
@@ -0,0 +1,63 @@
+        self.assertIsInstance(result, ResourceListOfInstrumentIdTypeDescriptor)
+        self.assertGreater(len(result.values), 0)
+
+    def test_get_api_with_access_token_env_var(self):
```
mock out the env vars here (this test has an assumption they have been set), see https://github.com/finbourne/lusid-sdk-python-preview/blob/master/sdk/tests/utilities/test_api_client_builder.py#L211 for an example on how to do this

rizwansaeed commented on sdk/tests/utilities/test_api_client_factory_auth_flow.py:
```diff
@@ -0,0 +1,63 @@
+source_config_details, config_keys = CredentialsSource.fetch_credentials(), CredentialsSource.fetch_config_keys()
+
+
+class ApiFactory(unittest.TestCase):
```
for all these tests we need to explicitly clear and set the env vars,  see https://github.com/finbourne/lusid-sdk-python-preview/blob/master/sdk/tests/utilities/test_api_client_builder.py#L211 for an example on how to do this

slemasne replied:
Sure, that is now completed.

slemasne commented on sdk/lusid/utilities/api_configuration.py:
```diff
@@ -4,7 +4,7 @@
 
     def __init__(self, token_url=None, api_url=None, username=None, password=None, client_id=None, client_secret=None,
-                 app_name=None, certificate_filename=None, proxy_config=None):
+                 app_name=None, certificate_filename=None, proxy_config=None, access_token=None):
```
Here we are adding the **access_token** as an attribute on the ApiConfiguration. 
This attribute gets loaded from the env vars.

slemasne commented on sdk/tests/utilities/test_api_client_builder.py:
```diff
@@ -44,8 +44,10 @@ def test_missing_from_config_file_throws(self, _, missing_attributes, token):
         api_config_kwargs["proxy_config"] = proxy_config
         api_configuration = ApiConfiguration(**api_config_kwargs)
+        api_configuration.access_token=None
 
```
The pat access token is now loaded into the **api_configuration**, so we need to excluded the pat token from the test.

slemasne commented on sdk/tests/utilities/test_api_client_builder.py:
```diff
@@ -185,6 +190,7 @@ def test_use_okta_response_handler(self):
         })
 
+        api_configuration.access_token = None
         api_configuration.certificate_filename = None
```
Again, we need to exclude the new pat from the test.

slemasne replied:
😃 
This is now fixed.
The previous change no longer required under revised approach.

slemasne commented on sdk/tests/utilities/test_api_client_builder.py:
```diff
@@ -195,16 +201,17 @@ def response_handler(okta_response):
 
         with patch.dict('os.environ', env_vars, clear=True), patch("requests.post") as mock_requests, self.assertRaises(
                 ValueError):
+
```
Now we need to mock the token **_after_** initializing the client, otherwise the client will load the pat from the env vars.

rizwansaeed replied:
this smells like we need to revisit the test setup so it is more self contained ie. the PAT should only exist if its being used in the test

slemasne replied:
This is now fixed.
The previous change no longer required under revised approach.

slemasne commented on sdk/tests/utilities/test_api_client_factory.py:
```diff
@@ -64,15 +74,20 @@ def test_get_unknown_api_throws_exception(self, _, api_to_build, error_message):
+
+        env_vars_wout_pat = self.get_env_vars_wout_pat()
+
+        with patch.dict(self.os_environ_dict_str, env_vars_wout_pat, clear=True):
```
Exclude the pat from the env vars

rizwansaeed commented on sdk/tests/utilities/test_api_client_factory.py:
```diff
@@ -44,6 +44,16 @@ def __getattribute__(self, name):
+    def setUpClass(cls):
+        cls.os_environ_dict_str = 'os.environ'
+
+    def get_env_vars_wout_pat(self):
```
wout? i presume you mean "without"?

slemasne replied:
Updated.

rizwansaeed replied:
did you accidentally miss a commit? latest version still has `wout`

rizwansaeed commented on sdk/tests/utilities/test_api_client_factory_auth_flow.py:
```diff
@@ -0,0 +1,119 @@
+                factory = ApiClientFactory(api_secrets_filename="INVALID_SECRETS")
+                api = factory.build(InstrumentsApi)
+
+                self.assertEqual(
```
should this assert on the good access token?

rizwansaeed replied:
ok i think i misunderstood the test, is this when the secrets doesn't exist rather than a bad value in it?

rizwansaeed commented on sdk/tests/utilities/test_api_client_factory_auth_flow.py:
```diff
@@ -0,0 +1,119 @@
+    def get_pat_env_var(self):
+
+        pat_env_vars = {
+            "FBN_LUSID_API_URL": os.getenv("FBN_LUSID_API_URL"),
```
these will cause tests to fail (locally) if using a secrets file, might need to do something like https://github.com/finbourne/lusid-sdk-python-preview/blob/master/sdk/tests/utilities/credentials_source.py#L31

slemasne replied:
Agreed. Updated.

slemasne commented on sdk/lusid/utilities/api_configuration_loader.py:
```diff
@@ -29,32 +30,55 @@ def load(api_secrets_filename=None):
+                for key, value in config_keys.items() if "proxy" not in key
+            }
+
+            if token is not None:
```
Agreed. Updated.

slemasne commented on sdk/tests/utilities/credentials_source.py:
```diff
@@ -38,7 +39,8 @@ def fetch_credentials(cls):
             "client_secret": os.getenv("FBN_CLIENT_SECRET", None),
-            "api_url": os.getenv("FBN_LUSID_API_URL", None)
+            "api_url": os.getenv("FBN_LUSID_API_URL", None),
+            "access_token": access_token
```
Agreed. Updated.

slemasne commented on sdk/tests/utilities/test_api_client_builder.py:
```diff
@@ -109,6 +111,9 @@ def test_build_client_no_token_provided_file_only(self):
         }
 
+        # Remove the access token from the secrets
+        del secrets["api"]["accessToken"]
```
Agreed. Updated.

slemasne commented on sdk/tests/utilities/test_api_client_builder.py:
```diff
@@ -185,6 +190,7 @@ def test_use_okta_response_handler(self):
             key: value for key, value in source_config_details.items() if "proxy" not in key
         })
 
+        api_configuration.access_token = None
```
Agreed. Updated.

rizwansaeed commented on sdk/lusid/utilities/api_client_builder.py:
```diff
@@ -52,17 +52,27 @@ def build(cls, api_secrets_filename=None, id_provider_response_handler=None, api
         # Load the configuration
         configuration = ApiConfigurationLoader.load(api_secrets_filename)
 
+        # If user passes a secrets file, then ignore the PAT loaded from env var
```
i don't think you need this check, can't you just set the local `api_token` based on whether one is based in or not?

rizwansaeed commented on sdk/tests/utilities/credentials_source.py:
```diff
@@ -60,8 +64,8 @@ def fetch_credentials(cls):
 
-        if None in vars.values():
-            assert False, "Source test configuration missing values from both secrets file and environment variables"
+        # if access_token is not None and None in vars.values():
```
why was this removed, looks like it is still needed

slemasne replied:
Agreed. Updated.

slemasne commented:
Re the final two comments:

**_did you accidentally miss a commit? latest version still has wout_**
Yes, updated.

_**ok i think i misunderstood the test, is this when the secrets doesn't exist rather than a bad value in it?**_
I updated file name to make intention clearer

rizwansaeed commented on sdk/lusid/utilities/api_client_builder.py:
```diff
@@ -58,11 +58,27 @@ def build(cls, api_secrets_filename=None, id_provider_response_handler=None, api
+        # The ApiClientBuilder uses the following conditional logic to resolve
+        # an authentication approach:
+        # (1) First look for a PAT passed in as a param by the user
+        # (2) Next look for a secrets file passed in as a param by the user
```
i personally find this a bit unclear, can we simplify 

1) Use PAT passed into the function
2) Use secrets files passed into the function
3) Use environment variables

PAT are used in preference to supplied credentials
secrets files are used in preference to environment variables

slemasne replied:
Done.

rizwansaeed commented:
can we fix up the code smells too
